src/ir/possible-contents.h

```diff
@@ -0,0 +1,577 @@
+  struct GlobalInfo {
+    Name name;
+    // The type of the global in the module. We stash this here so that we do
+    // not need to pass around a module all the time.
+    // TODO: could we save size in this variant if we did pass around the
+    //       module?
+    Type type;
+    bool operator==(const GlobalInfo& other) const {
+      return name == other.name && type == other.type;
+    }
+  };
+
+  using ExactType = Type;
+
+  struct Many : public std::monostate {};
+
+  // TODO: This is similar to the variant in PossibleConstantValues, and perhaps
+  //       we could share code, but extending a variant using template magic may
+  //       not be worthwhile. Another option might be to make PCV inherit from
+  //       this and disallow ExactType etc., but PCV might get slower.
+  using Variant = std::variant<None, Literal, GlobalInfo, ExactType, Many>;
+  Variant value;
+
+public:
+  PossibleContents() : value(None()) {}
+  PossibleContents(const PossibleContents& other) : value(other.value) {}
+
+  // Most users will use one of the following static functions to construct a
+  // new instance:
+
+  static PossibleContents none() {
+    PossibleContents ret;
+    ret.value = None();
+    return ret;
+  }
+  static PossibleContents literal(Literal c) {
+    PossibleContents ret;
+    ret.value = c;
+    return ret;
+  }
+  static PossibleContents global(Name name, Type type) {
+    PossibleContents ret;
+    ret.value = GlobalInfo{name, type};
+    return ret;
+  }
+  static PossibleContents exactType(Type type) {
+    PossibleContents ret;
+    ret.value = ExactType(type);
+    return ret;
+  }
+  static PossibleContents many() {
+    PossibleContents ret;
+    ret.value = Many();
+    return ret;
+  }
+
+  PossibleContents& operator=(const PossibleContents& other) {
+    value = other.value;
+    return *this;
+  }
+
+  bool operator==(const PossibleContents& other) const {
+    return value == other.value;
+  }
+
+  bool operator!=(const PossibleContents& other) const {
+    return !(*this == other);
+  }
+
+  // Combine the information in a given PossibleContents to this one. The
+  // contents here will then include whatever content was possible in |other|.
+  void combine(const PossibleContents& other) {
+    // First handle the trivial cases of them being equal, or one of them is
+    // None or Many.
+    if (*this == other) {
+      return;
+    }
+    if (other.isNone()) {
+      return;
+    }
+    if (isNone()) {
+      value = other.value;
+      return;
+    }
+    if (isMany()) {
+      return;
+    }
+    if (other.isMany()) {
+      value = Many();
+      return;
+    }
+
+    auto type = getType();
+    auto otherType = other.getType();
+
+    if (!type.isRef() || !otherType.isRef()) {
+      // At least one is not a reference. We could in principle try to find
+      // ExactType here, say as the combination of two different constants, but
+      // as subtyping does not exist, Many is good enough anyhow, so do that.
```

Comment:
```suggestion
      // ExactType here, say as the combination of two different constants, but
      // since there is no subtyping between non-ref types, ExactType would not carry any more information than Many.
```

Alternatively, it seems like it would be pretty harmless to get the shared ExactType in the case of two literals, and that might be less surprising to readers.